### PR TITLE
[MLIR] Disallow `MeasureOp`, `PythonCallOp` and `CustomCallOp` in gradients.

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:** MeasureOp, PythonCallOp and CustomCallOp are allowed in gradients even though we don't support them yet. This leads to logic errors and bad UI/UX.

**Description of the Change:** Verify that these operations are not reachable from the `grad` op.

**Benefits:** Better UI/UX.

[sc-42761]

### CODE REVIEW

**Review by josh146 — COMMENTED** [review]
Thanks @erick-xanadu! I'm curious on the underlying issue - is it because `measure` is returning an `int`?

**Comment by erick-xanadu** [comment]
It is because the measurement is non-deterministic. If I'm wrong about this I can close this PR. I don't believe we currently handle the case for differentiating the following function in general.

```python
def f(x):
  qml.RX(x, wires=0)
  return measure(0)
```

**Comment by dime10** [comment]
This is tricky, because you can use them with finite-diff if the overall function is deterministic. For example:

```py
def f(x):
    qml.RY(x, wires=0)
    m = measure(0)
    cond(m)(lambda: qml.PauliX(0) and None)()
    return qml.expval(qml.PauliZ(0))
```

But even a quantum function with 

``py
def f(x):
    qml.RY(x, wires=0)
    return measure(0)
```

that is post-processed into a deterministic output somehow could also be differentiable. 

Having said that, I could see for UX reasons that we might want to disallow those case to prevent misuse on non-differentiable functions. But I'm not yet convinced that is the right move.

File: frontend/test/pytest/test_jax_integration.py [hunk 1]
```diff
@@ -346,8 +346,8 @@ def cost_fn(x, y):
 
         result1 = cost_fn(0.0, 0.0)
         result2 = cost_fn(0.0, jnp.pi)
-        assert jnp.allclose(result1, 0.0)
-        assert jnp.allclose(result2, 0.0)
+        assert jnp.allclose(result1, 0.0, atol=1e-7)
+        assert jnp.allclose(result2, 0.0, atol=1e-7)
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_jax_integration.py:350]: I think this was just a rebase error of mine.

```suggestion
        assert jnp.allclose(result1, 0.0)
        assert jnp.allclose(result2, 0.0)
```

File: frontend/test/pytest/test_gradient.py [hunk 2]
```diff
@@ -1172,5 +1175,41 @@ def interpreted(x):
     assert np.allclose(compiled(inp), interpreted(inp))
 
 
+class TestGradientErrors:
```
Comment (dime10) [inline frontend/test/pytest/test_gradient.py:1178]: The changelog also mentions that zne would raise an error, should add a test for this as well :)

File: mlir/lib/Quantum/Transforms/annotate_function.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+{
+    CallGraphNode *callee = edge.getTarget();
+    return getFuncOp(callee, cg);
+}
+
+bool anyCalleeIsAnnotated(func::FuncOp op, const char *attr, CallGraph &cg)
+{
+    Region &region = op.getRegion();
+    CallGraphNode *node = cg.lookupNode(&region);
+    // TODO: ICE if we do not find the node.
```
Comment (dime10) [inline mlir/lib/Quantum/Transforms/annotate_function.cpp:100]: What's ICE?

File: mlir/lib/Quantum/Transforms/annotate_function.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    for (auto i = node->begin(), e = node->end(); i != e; ++i) {
+        std::optional<func::FuncOp> maybeCallee = getCallee(*i, cg);
+        // An indirect call
+        if (!maybeCallee)
+            return true;
```
Comment (dime10) [inline mlir/lib/Quantum/Transforms/annotate_function.cpp:105]: What is this case and why is it true?

File: mlir/lib/Quantum/Transforms/annotate_function.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+{
+    return !isAnnotated(op, hasMeasureAttrName) && invalidGradientOperation(op);
+}
+
+void annotate(func::FuncOp op, PatternRewriter &rewriter, const char *attr)
+{
+    op->setAttr(attr, rewriter.getUnitAttr());
+}
+
+struct AnnotateFunctionTransform : public OpRewritePattern<func::FuncOp> {
```
Comment (dime10) [inline mlir/lib/Quantum/Transforms/annotate_function.cpp:61]: Can you call these patterns instead of transforms (in the class name)? Easier to read when sticking with the MLIR coding conventions.

**Review by dime10 — COMMENTED** [review]
Thanks @erick-xanadu, looks great!

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 3 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp:100]: Internal Compiler Error.

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 4 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp:105]: For indirect function calls, you don't know the target of the call. This doesn't happen for us. But it is safer to say that if there are indirect function calls which cannot be resolved then they could very likely contain these operations.

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 5 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: https://github.com/PennyLaneAI/catalyst/pull/204/commits/4978a063f00842268cf10f89568e91c59553f663

File: frontend/test/pytest/test_gradient.py (hunk 2 above)
Comment (erick-xanadu) [reply frontend/test/pytest/test_gradient.py:1179]: https://github.com/PennyLaneAI/catalyst/pull/204/commits/1390d23e18fb798ff762bdcbd41d10ba3fd710d7

**Comment by erick-xanadu** [comment]
@josh146 (or anyone :sweat_smile: ) can you provide a hand and move this shortcut ticket [sc-42761] to the current quarter / iteration?

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 4 above)
Comment (dime10) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp:105]: That makes sense, could you state that in the comment?

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 3 above)
Comment (dime10) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp:100]: Should we add that now?

Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp:100]: Mmm... I am always confused about which is the correct API to state this to the user and then end the compilation process. I don't think it is singalPassFailure but it might be.

Comment (dime10) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp:100]: I think `signalPassFailure` indicates an unrecoverable state, which will most likely abort compilation entirely. If it is case that the pass simply doesn't apply and should ignore, then a match failure would be better. If it is a case that we don't expect though, and shouldn't exist, than the pass failure is appropriate I think :)

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 4 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: I added a comment along with a paper for devirtualization methods with different levels of precision. I doubt any of these analyses is already implemented in MLIR, but it is a good summary on the topic. https://github.com/PennyLaneAI/catalyst/pull/204/commits/ab2518c6207692ee27164463da901d4709de2bd8

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 3 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: We are already using `signalPassFailure` but it can only be used in the `Pass` class, not in the `Pattern` class, which is where the transformation happens. Essentially, what I would like to do is exit the pattern early and signal failure, but that is only possible through the `LogicalResult` which implies I catch the error during the `match` but not `rewrite` portion of the pattern. I think leaving it as is, is ok-ish. It would be better to emit a message and then abort. This is what I think `quantum-opt` should do, but not what the compiler driver should do. The compiler driver should likely raise an exception (?)

Comment (dime10) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: In that case an exception is also a reasonable choice 👍

Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: Mmm... but this file is compiled with `-no-exceptions` :sweat_smile: 

```
/home/ubuntu/code/catalyst/mlir/lib/Quantum/Transforms/annotate_function.cpp:103:9: error: cannot use 'throw' with exceptions disabled
        throw std::runtime_error("Internal Compiler Error: This is a state that should not happen");
        ^
```

Comment (dime10) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: Oh interesting. Assertions do work, but unfortunately they don't produce an error message in Python and instead just crash the process :/

One easy thing you can do to go with your above plan is to combine match and rewrite into the same step (`matchAndRewrite`), then you can also return a logical result from the "rewrite" step.
But I don't see how that would signal an error state, because "failure" just match the pattern didn't match which is not necessarily an error state 🤔

Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: I think it is fine to leave as is (or have an assertion an abort). This case would signal an error in MLIR's infrastructure.

File: mlir/lib/Gradient/IR/GradientOps.cpp [hunk 6]
```diff
@@ -54,6 +55,15 @@ LogicalResult verifyGradInputs(OpState *op_state, func::FuncOp callee, ValueRang
         return op_state->emitOpError("incorrect number of operands for callee, ")
                << "expected " << fnType.getNumInputs() << " but got " << fnArgs.size();
 
+    if (callee->getAttrOfType<UnitAttr>("catalyst.invalidGradientOperation")) {
```
Comment (grwlf) [inline mlir/lib/Gradient/IR/GradientOps.cpp:58]: Is there a way to register this attribute somewhere in our dialect? e.g. [here](https://github.com/PennyLaneAI/catalyst/blob/663de3c89c25c7d949ce0f8fea9578c15abcd1cf/mlir/include/Quantum/IR/QuantumOps.td#L64)

File: mlir/lib/Quantum/Transforms/annotate_function.cpp [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+#include "Catalyst/IR/CatalystOps.h"
+#include "Quantum/Transforms/Passes.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+static constexpr const char *hasMeasureAttrName = "catalyst.invalidGradientOperation";
```
Comment (grwlf) [inline mlir/lib/Quantum/Transforms/annotate_function.cpp:30]: 1. If we can't encode it in the *td files, let me suggest to move this string constant to some header to only declare it once.
2. Wouldn't it better to name this constant `invalidGradientOperationAttrName` ?

File: mlir/lib/Gradient/IR/GradientOps.cpp (hunk 6 above)
Comment (erick-xanadu) [reply mlir/lib/Gradient/IR/GradientOps.cpp:58]: Yes! Potentially. I am not an expert yet on using MLIR's libraries. :sweat_smile: I'll give it a try.

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 3 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: Ok, I just deleted the comment based on our discussion.

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 7 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: https://github.com/PennyLaneAI/catalyst/pull/204/commits/c1a8915f44ee3fc1b6e21ba8a40b8b44298ee3ba https://github.com/PennyLaneAI/catalyst/pull/204/commits/4756e695fa567b96516dc0488f714878e51d83a3

File: mlir/lib/Quantum/Transforms/annotate_function.cpp (hunk 3 above)
Comment (erick-xanadu) [reply mlir/lib/Quantum/Transforms/annotate_function.cpp]: Added the assertion https://github.com/PennyLaneAI/catalyst/pull/204/commits/0271f1e25e9d129abb0609d02843824d10bb5051

**Review by dime10 — APPROVED** [review]
